### PR TITLE
Tracing atomic operations

### PR DESCRIPTION
**Pull request by SokolovaMaria**
Fixes #19
Supported internal `Trace` fields. 
Added format function to the `Trace` constructor. Now declaration may look like:
```private val t = trace { i, t -> "..." }```
Supported trace erasing from JS.

### CODE REVIEW

File: atomicfu-common/src/main/kotlin/kotlinx/atomicfu/Trace.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+fun trace(size: Int = 32, format: (AtomicInt, String) -> String = { index, text -> "$index: $text" }): Trace = TraceImpl(size, format)
+
+val NO_TRACE = Trace()
+
+/**
+ * Default no-op Trace implementation that can be overridden
+ */
+public open class Trace {
+
+    @JsName("atomicfu\$trace\$append\$")
```
Comment (SokolovaMaria) [inline atomicfu-common/src/main/kotlin/kotlinx/atomicfu/Trace.kt]: Seems, that I should create ```expected interface Trace {..}``` , then actual declarations of `Trace` in modules for every platform and move these @JsName annotations to JS module (to support Trace on Native too).
Is there a way to add JS annotations to common code?

File: README.md [hunk 2]
```diff
@@ -330,3 +330,30 @@ execution of tests with the original (non-transformed) classes for Maven:
 ```
 
 For Gradle there is nothing else to add. Tests are always run using original (non-transformed) classes.
+
+### Multiplatform debug using atomicfu trace
+
+You can debug your tests tracing atomic operations with a special trace object:
+```kotlin
+private val t = trace()
```
Comment (elizarov) [inline README.md]: The "constructor" for trace shall be named `Trace()` (with capital T)

File: atomicfu-common/src/main/kotlin/kotlinx/atomicfu/Trace.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * Usage: create a separate field for trace and pass to atomic factory function:
+ *
+ * ```
+ * val trace = trace(size)
+ * val a = atomic(initialValue, trace)
+ * ```
+ */
+@JsName("atomicfu\$trace\$")
+fun trace(size: Int = 32, format: (AtomicInt, String) -> String = { index, text -> "$index: $text" }): Trace = TraceImpl(size, format)
```
Comment (elizarov) [inline atomicfu-common/src/main/kotlin/kotlinx/atomicfu/Trace.kt]: The "constructor" for trace shall be named `Trace()` (with capital T)

File: atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/FlowAnalyzer.kt [hunk 4]
```diff
@@ -19,9 +19,10 @@ package kotlinx.atomicfu.transformer
 import org.objectweb.asm.Opcodes.*
 import org.objectweb.asm.Type
 import org.objectweb.asm.tree.*
+import kotlin.math.abs
 
 class FlowAnalyzer(
-    private val start: AbstractInsnNode?
+        private val start: AbstractInsnNode?
```
Comment (elizarov) [inline atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/FlowAnalyzer.kt]: Extra tab level (remove)

File: atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt [hunk 5]
```diff
@@ -600,6 +659,7 @@ class AtomicFUTransformer(
                             val next = i.nextUseful
                             val fieldId = (next as? FieldInsnNode)?.checkPutField()
                                 ?: abort("factory $methodId invocation must be followed by putfield")
+                            eraseTraceInit(i.previous)
```
Comment (elizarov) [inline atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt]: It does not work. It attempts to `eraseTraceInit` even if there is no trace parameters in this constructor invocation. It fails on this code:

```
class Counter {
    private val a = atomic(0)
    private val trace = trace(64)

    fun inc(): Int {
        trace { "inc() invoked" }
        val x = a.incrementAndGet()
        trace { "inc() = $x" }
        return x
    }

    internal fun get() = a.value
}
``` 

with

```
Trace.kt: kotlinx.coroutines.experimental.channels.Trace::<init>: Failed to transform: java.lang.ClassCastException: org.objectweb.asm.tree.InsnNode cannot be cast to org.objectweb.asm.tree.MethodInsnNode
java.lang.ClassCastException: org.objectweb.asm.tree.InsnNode cannot be cast to org.objectweb.asm.tree.MethodInsnNode
        at kotlinx.atomicfu.transformer.FlowAnalyzer.getInitStart(FlowAnalyzer.kt:64)
        at kotlinx.atomicfu.transformer.AtomicFUTransformer$TransformerMV.eraseTraceInit(AtomicFUTransformer.kt:619)
        at kotlinx.atomicfu.transformer.AtomicFUTransformer$TransformerMV.transform(AtomicFUTransformer.kt:662)
        at kotlinx.atomicfu.transformer.AtomicFUTransformer$TransformerMV.visitEnd(AtomicFUTransformer.kt:470)
```

**Comment by SokolovaMaria** [comment]
Superceded by #114

**Pull request closed without merging**